docs/features/non-source-outputs.md

```diff
@@ -0,0 +1,141 @@
+
+## Open Questions
+
+- Naming: This proposal originally used the term `NonSource` but that's not particularly descriptive and perhaps odd in that it's not actually the mirror of a source output in that it's not going into the compilation. `Artifact` seems like a better term? Or simply `File`?
+- Do we want custom mapping for the `compiler_artifact.` scheme, or could we just use the existing `build_metadata.<ItemType>.<Metadata>` mechanism and recognize `CompilerMappedArtifact` as a special key?
+  - This is doable, but does tie the MSBuild item type name directly into the compiler, rather than having a level of indirection.
+- File mapping: should we just let the generators party on the file system and allow absolute paths and `.`, `..` etc, rather than having the host control the output location?
+  - An issue with this approach is that a user can control their project layout via MSBuild, and the generator won't know that it has changed, meaning you might end up with files in odd places
+  - It would be possible to route though e.g. `OutDir` either automatically, or by the generator author specifying it as a `CompilerVisibleProperty` but this relies on the author doing the correct thing (and multiple authors having to copy the same logic)
+  - It seems better to write to a 'virtual' location, and let the host re-direct that as appropriate.
+- File mapping: within a hint name should we allow sub-directories? So that a given generator can write to e.g. `<base>/css` and `<base>/js` and just map the `<base>` directory to wherever is appropriate?
+- File mapping: can we just have the compiler write them to the obj directory (or the `GeneratedFilesOut` directory) and use MSBuild to do the copying? Either by default or require the author to supply targets to do the copying after build?
+  - This makes things for the compiler simpler, but doesn't optimize for disk hits or consistency. A given generator may not even _know_ what its outputs are going to be ahead of time, as the artifact name may be derived from a user input, making it almost impossible for the generator to provide the copy
+    - We could say that each generators outputs go to a specific folder, like the generated files do, so that it could just copy `<dir>/*` to the location that matters.
+- File mapping: do we really need the ability to map individual artifacts? Maybe per-generator is enough, and a user can always write `MSBuild` logic to move the individual artifacts if required?
+- User control: do we need a way for the user to disable artifact production? Should this be per generator or just all artifacts?
```

Comment:
Thinking more, I think this is definitely needed.

Reply:
What scenarios do you think this is needed for? 

The one that popped to mind for me was WPF temporary assemblies. Having artifact production there is not desired.

Reply:
Yeah, or just users doing weird stuff with builds, which they very much do :) I worry about not having an escape hatch at all. Could be as simple as we don't emit unless the output path is passed to the compiler (that's what we do with the source files too). So, by default, it always is on, but you can disable it via MSBuild trivially for whatever reason you have.